### ISSUE TITLE
Support serializable sealed interfaces

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.serialization#1417]
With support for the `sealed interface` in Kotlin `1.5.0`, it would be extremely useful to support `@Serializable sealed interface`, making manual polymorphic registration of subclasses unnecessary.

**Comment by rgbrizzlehizzle** [comment]
For people looking for a work around, I have put a working solution together using `sealed interface` and `@JvmInline`

For an API client I am working on, I can declare a `sealed interface`

```kotlin
@Suppress("VariableNaming")
@Serializable(with = UserSerializer::class)
sealed interface User {
  val `object`: String
  val id: UUID
  val type: UserType?
  val name: String?
  val avatarUrl: String?
}
```

add my implementations, for example

```kotlin
@Serializable
data class Person(
  override val `object`: String,
  @Serializable(with = UUIDSerializer::class)
  override val id: UUID,
  override val type: UserType?,
  override val name: String?,
  @SerialName("avatar_url")
  override val avatarUrl: String?,
  val person: PersonInfo
) : User

@Serializable
data class PersonInfo(
  val email: String
)
```

my serializer 

```kotlin
object UserSerializer : JsonContentPolymorphicSerializer<User>(User::class) {
  override fun selectDeserializer(element: JsonElement) = when {
    "person" in element.jsonObject -> Person.serializer()
    "bot" in element.jsonObject -> Bot.serializer()
    else -> error("Unhandled user type in $element")
  }
}
```

and finally, add a `JvmInline` class

```kotlin
@JvmInline
@Serializable
value class UserResponse(val user: User)
```

At which point I can successfully send a ktor request

```kotlin
suspend fun getUser(userId: String): UserResponse = klient.get("$baseUrl/users/$userId")
```

and get my payload 

```
UserResponse(user=Person(object=user, id=omitted, type=PERSON, name=omitted, avatarUrl=null, person=PersonInfo(email=omitted)))
```

Hope this helps 😃 

Also worth noting, wrapping in a generic like a `List<User>` works as well!

**Comment by williamboxhall** [comment]
I'd love a proper fix for this one too please :) as mentioned in the linked issue, I'm trying to use sealed interfaces Avro4k which builds on top of kotlinx-serialization. `sealed interface` are coming back as `SerialKind.OPEN` instead of `SerialKind.SEALED`

**Comment by sandwwraith** [comment]
Fixed in #1576 (Kotlin 1.6.20)
